Client/Frontend/Browser/TopTabsViewController.swift

```diff
@@ -0,0 +1,714 @@
+        if let frame = collectionView.layoutAttributesForItemAtIndexPath(NSIndexPath(forRow: index, inSection: 0))?.frame {
+            // Padding is added to ensure the tab is completely visible (none of the tab is under the fader)
+            let padFrame = frame.insetBy(dx: -(TopTabsUX.TopTabsBackgroundShadowWidth+TopTabsUX.FaderPading), dy: 0)
+            collectionView.scrollRectToVisible(padFrame, animated: true)
+        }
+    }
+    
+}
+
+extension TopTabsViewController: Themeable {
+    func applyTheme(themeName: String) {
+        tabsButton.applyTheme(themeName)
+        if themeName == Theme.PrivateMode {
+            isPrivate = true
+        }
+        else if themeName == Theme.NormalMode {
+            isPrivate = false
+        }
+    }
+}
+
+extension TopTabsViewController: TopTabCellDelegate {
+    func tabCellDidClose(cell: TopTabCell) {
+        let indexPath = collectionView.indexPathForCell(cell)!
+        let tab = tabsToDisplay[indexPath.item]
+        if tab == tabManager.selectedTab {
+            delegate.topTabsDidChangeTab()
+        }
+        if tabsToDisplay.count == 1 {
+            tabManager.removeTab(tab)
+            tabManager.selectTab(tabsToDisplay.first)
+            collectionView.reloadData()
+        }
+        else {
+            tabManager.removeTab(tab)
+            self.collectionView.performBatchUpdates({
+                self.collectionView.deleteItemsAtIndexPaths([indexPath])
+                }, completion: { finished in
+                    if #available(iOS 9, *) {
+                        self.collectionView.reloadData()
+                    }
+            })
+        }
+    }
+}
+
+extension TopTabsViewController: UICollectionViewDataSource {
+    @objc func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
+        let index = indexPath.item
+        let tabCell = collectionView.dequeueReusableCellWithReuseIdentifier(TopTabCell.Identifier, forIndexPath: indexPath) as! TopTabCell
+        tabCell.delegate = self
+        
+        let tab = tabsToDisplay[index]
+        tabCell.style = tab.isPrivate ? .Dark : .Light
+        tabCell.titleText.text = tab.displayTitle
+        
+        if tab.displayTitle.isEmpty {
+            if (tab.url?.baseDomain()?.contains("localhost") ?? false || tab.url == nil) {
+                tabCell.titleText.text = Strings.TopTabsNewTabTitle
+            }
+            else {
+                tabCell.titleText.text = tab.displayURL?.absoluteString
+            }
+        }
+        
+        tabCell.selectedTab = tab == tabManager.selectedTab
+        
+        if index > 0 && index < tabsToDisplay.count && tabsToDisplay[index] != tabManager.selectedTab && tabsToDisplay[index-1] != tabManager.selectedTab {
+            tabCell.seperatorLine = true
+        }
+        else {
+            tabCell.seperatorLine = false
+        }
+        
+        if !tab.displayTitle.isEmpty {
```

Comment:
Nit: You already have an `if` statement above checking the inverse of this boolean. Maybe combine the two?